pkg/loop/internal/capabilities.go

```diff
@@ -0,0 +1,587 @@
+	})
+	return &emptypb.Empty{}, err
+}
+
+func (c *callbackExecutableServer) UnregisterFromWorkflow(ctx context.Context, req *pb.UnregisterFromWorkflowRequest) (*emptypb.Empty, error) {
+	config, err := values.FromProto(req.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.impl.UnregisterFromWorkflow(ctx, capabilities.UnregisterFromWorkflowRequest{
+		Metadata: capabilities.RegistrationMetadata{
+			WorkflowID: req.Metadata.WorkflowId,
+		},
+		Config: config.(*values.Map),
+	})
+	return &emptypb.Empty{}, err
+}
+
+func (c *callbackExecutableServer) Execute(ctx context.Context, req *pb.ExecuteRequest) (*emptypb.Empty, error) {
+	ch := make(chan capabilities.CapabilityResponse)
+
+	conn, err := c.dial(req.CallbackId)
+	if err != nil {
+		return nil, err
+	}
+
+	connCtx, connCancel := context.WithCancel(context.Background())
+	go callbackIssuer(connCtx, pb.NewCallbackClient(conn), ch, c.Logger)
+
+	cr := req.CapabilityRequest
+	md := cr.Metadata
+
+	config, err := values.FromProto(cr.Config)
+	if err != nil {
+		connCancel()
+		return nil, err
+	}
+
+	inputs, err := values.FromProto(cr.Inputs)
+	if err != nil {
+		connCancel()
+		return nil, err
+	}
+
+	r := capabilities.CapabilityRequest{
+		Metadata: capabilities.RequestMetadata{
+			WorkflowID:          md.WorkflowId,
+			WorkflowExecutionID: md.WorkflowExecutionId,
+		},
+		Config: config.(*values.Map),
+		Inputs: inputs.(*values.Map),
+	}
+
+	err = c.impl.Execute(ctx, ch, r)
+	if err != nil {
+		connCancel()
+		return nil, err
+	}
+
+	c.cancelFuncs[md.WorkflowId] = connCancel
+	return &emptypb.Empty{}, nil
+}
+
+type callbackExecutableClient struct {
+	grpc pb.CallbackExecutableClient
+	*brokerExt
+}
+
+func newCallbackExecutableClient(brokerExt *brokerExt, conn *grpc.ClientConn) *callbackExecutableClient {
+	return &callbackExecutableClient{
+		grpc:      pb.NewCallbackExecutableClient(conn),
+		brokerExt: brokerExt,
+	}
+}
+
+var _ capabilities.CallbackExecutable = (*callbackExecutableClient)(nil)
+
+func toProto(req capabilities.CapabilityRequest) (*pb.CapabilityRequest, error) {
+	inputs := &values.Map{Underlying: map[string]values.Value{}}
+	if req.Inputs != nil {
+		inputs = req.Inputs
+	}
+
+	inputsPb, err := inputs.Proto()
+	if err != nil {
+		return nil, err
+	}
+
+	config := &values.Map{Underlying: map[string]values.Value{}}
+	if req.Config != nil {
+		config = req.Config
+	}
+
+	configPb, err := config.Proto()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.CapabilityRequest{
+		Metadata: &pb.RequestMetadata{
+			WorkflowId:          req.Metadata.WorkflowID,
+			WorkflowExecutionId: req.Metadata.WorkflowExecutionID,
+		},
+		Inputs: inputsPb,
+		Config: configPb,
+	}, nil
+}
+
+func (c *callbackExecutableClient) Execute(ctx context.Context, callback chan<- capabilities.CapabilityResponse, req capabilities.CapabilityRequest) error {
+	cid, res, err := c.serveNew("Callback", func(s *grpc.Server) {
+		pb.RegisterCallbackServer(s, newCallbackServer(callback))
```

Comment:
Same question as on the previous PR - why do we need to create a new server for each call rather than re-using a single one per capability?

Reply:
OK, I did a bit of digging on this: under the hood what this does is just register a handler to a server, so it should be fairly inexpensive. However, we can't be guaranteed to receive the same underlying grpc.Server instance, so my original solution (to wrap the register call in a sync.Once) won't work.

With the understanding that this is fairly inexpensive and the expected way, are you happy to approve as is and we can revisit it if the overhead of these servers becomes a problem?